File: quant1x/std/filesystem.go
package std

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"sync"
)

// DisableCache will disable caching of the home directory. Caching is enabled
// by default.
var DisableCache bool

var homedirCache string
var cacheLock sync.RWMutex

// HomeDir returns the home directory for the executing user.
//
// This uses an OS-specific method for discovering the home directory.
// An error is returned if a home directory cannot be detected.
func HomeDir() (string, error) {
	if !DisableCache {
		cacheLock.RLock()
		cached := homedirCache
		cacheLock.RUnlock()
		if cached != "" {
			return cached, nil
		}
	}

	cacheLock.Lock()
	defer cacheLock.Unlock()

	var result string
	var err error
	if runtime.GOOS == "windows" {
		result, err = dirWindows()
	} else {
		// Unix-like system, so just assume Unix
		result, err = dirUnix()
	}

	if err != nil {
		return "", err
	}
	homedirCache = result
	return result, nil
}

// ExpandUser expands the path to include the home directory if the path
// is prefixed with `~`. If it isn't prefixed with `~`, the path is
// returned as-is.
func ExpandUser(path string) (string, error) {
	path = strings.TrimSpace(path)
	if len(path) == 0 {
		return path, nil
	}

	if path[0] != '~' {
		return path, nil
	}

	if len(path) > 1 && path[1] != '/' && path[1] != '\\' {
		return "", errors.New("cannot expand user-specific home dir")
	}

	dir, err := HomeDir()
	if err != nil {
		return "", err
	}

	if len(path) == 1 {
		return dir, nil
	}

	return filepath.Join(dir, path[2:]), nil
}

// Reset clears the cache, forcing the next call to HomeDir to re-detect
// the home directory. This generally never has to be called, but can be
// useful in tests if you're modifying the home directory via the HOME
// env var or something.
func Reset() {
	cacheLock.Lock()
	defer cacheLock.Unlock()
	homedirCache = ""
}

const (
	EnvQuant1xHome = "QUANT1X_HOME"
	EnvGoxHome     = "GOX_HOME"
)

func dirUnix() (string, error) {
	if home := os.Getenv(EnvQuant1xHome); home != "" {
		return home, nil
	}
	if home := os.Getenv(EnvGoxHome); home != "" {
		return home, nil
	}
	homeEnv := "HOME"
	if runtime.GOOS == "plan9" {
		// On plan9, env vars are lowercase.
		homeEnv = "home"
	}

	// First prefer the HOME environmental variable
	if home := os.Getenv(homeEnv); home != "" {
		return home, nil
	}

	var stdout bytes.Buffer

	// If that fails, try OS specific commands
	if runtime.GOOS == "darwin" {
		cmd := exec.Command("sh", "-c", `dscl -q . -read /Users/"$(whoami)" NFSHomeDirectory | sed 's/^[^ ]*: //'`)
		cmd.Stdout = &stdout
		if err := cmd.Run(); err == nil {
			result := strings.TrimSpace(stdout.String())
			if result != "" {
				return result, nil
			}
		}
	} else {
		cmd := exec.Command("getent", "passwd", strconv.Itoa(os.Getuid()))
		cmd.Stdout = &stdout
		if err := cmd.Run(); err != nil {
			// If the error is ErrNotFound, we ignore it. Otherwise, return it.
			if err != exec.ErrNotFound {
				return "", err
			}
		} else {
			if passwd := strings.TrimSpace(stdout.String()); passwd != "" {
				// username:password:uid:gid:gecos:home:shell
				passwdParts := strings.SplitN(passwd, ":", 7)
				if len(passwdParts) > 5 {
					return passwdParts[5], nil
				}
			}
		}
	}

	// If all else fails, try the shell
	stdout.Reset()
	cmd := exec.Command("sh", "-c", "cd && pwd")
	cmd.Stdout = &stdout
	if err := cmd.Run(); err != nil {
		return "", err
	}

	result := strings.TrimSpace(stdout.String())
	if result == "" {
		return "", errors.New("blank output when reading home directory")
	}

	return result, nil
}

func dirWindows() (string, error) {
	if home := os.Getenv(EnvQuant1xHome); home != "" {
		return home, nil
	}
	// 启用环境变量GOX_HOME是为了Windows服务以系统账户运行时无法获取登录用户的宿主目录而预备的
	if home := os.Getenv(EnvGoxHome); home != "" {
		return home, nil
	}
	// First prefer the HOME environmental variable
	if home := os.Getenv("HOME"); home != "" {
		return home, nil
	}

	// Prefer standard environment variable USERPROFILE
	if home := os.Getenv("USERPROFILE"); home != "" {
		return home, nil
	}

	drive := os.Getenv("HOMEDRIVE")
	path := os.Getenv("HOMEPATH")
	home := drive + path
	if drive == "" || path == "" {
		return "", errors.New("HOMEDRIVE, HOMEPATH, or USERPROFILE are blank")
	}

	return home, nil
}

const (
	CACHE_DIR_PERMS  = 0755 // 目录权限
	CACHE_FILE_PERMS = 0644 // 文件权限
)

// MkDirs 创建指定路径的所有目录（包括任何必要的父目录），使用默认的目录权限
func MkDirs(path_ string, notExistToCreate ...bool) error {
	path, err := ExpandUser(path_)
	if err != nil {
		return err
	}
	create := true
	if len(notExistToCreate) > 0 {
		create = notExistToCreate[0]
	}
	if create {
		return os.MkdirAll(path, CACHE_DIR_PERMS)
	}
	info, err := os.Stat(path)
	if err != nil {
		return err
	}
	if !info.IsDir() {
		return fmt.Errorf("path exists but is not a directory: %s", path)
	}
	return nil
}

// CheckFilepath
//
//	检查filename 文件路径, 如果不存在就创建
func CheckFilepath(path string, notExistToCreate ...bool) error {
	dir := filepath.Dir(path)
	return MkDirs(dir, notExistToCreate...)
}

// FileExist 路径是否存在
func FileExist(path string) bool {
	_, err := os.Lstat(path)
	return !os.IsNotExist(err)
}

File: quant1x/std/filesystem.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import os
import time
import tempfile


def homedir() -> str:
    """
    获取当前用户的主目录
    优先级:
    1. QUANT1X_HOME 环境变量
    2. GOX_HOME 环境变量
    3. HOME 环境变量
    4. USERPROFILE 环境变量 (Windows only)
    5. 系统临时目录 (fallback)
    """
    home = os.getenv("QUANT1X_HOME")
    if home:
        return home
    
    home = os.getenv("GOX_HOME")
    if home:
        return home
        
    home = os.getenv("HOME")
    if home:
        return home
        
    if os.name == 'nt':
        home = os.getenv("USERPROFILE")
        if home:
            return home
            
    return tempfile.gettempdir()

def expand_user(path: str) -> str:
    """
    展开用户主目录 (例如 "~/data" -> "/home/user/data")
    """
    if not path:
        return path
        
    path = path.strip()
    if not path.startswith('~'):
        return path
        
    home = homedir()
    if len(path) == 1:
        return home
        
    if path[1] == '/' or path[1] == '\\':
        return os.path.join(home, path[2:])
        
    return path

def mkdirs(path: str):
    """
    创建目录
    :param path:
    :return:
    """
    if not os.path.exists(path):
        os.makedirs(path)


def touch(filename: str):
    """
    创建一个空文件
    :param filename:
    :return:
    """
    directory = os.path.dirname(filename)
    mkdirs(directory)
    with open(filename, 'w') as done_file:
        pass

def update_file_mtime(filename: str, timestamp: float = 0):
    """
    更新文件的修改时间(mtime)，如果文件不存在则创建空文件
    
    Args:
        filename (str): 需要更新时间的文件路径
        timestamp (float): 指定的时间戳，默认为0表示使用当前时间
    
    Raises:
        Exception: 文件操作失败时静默处理（不抛出异常）
    """
    if timestamp == 0:
        timestamp = time.time()
    try:
        if not os.path.exists(filename):
            touch(filename)
        os.utime(filename, (int(timestamp), int(timestamp)))
    except Exception:
        pass